PYIC-6246: Cross-browser recovery restart journey

MikeCollingwood opened the pull request:
## Proposed changes

### What changed

- Add Cross-browser recovery restart journey logic to `CheckExistingIdentityHandler`
- Centralise Async CRI status checking with new `AsyncCriStatus` class

### Why did it change

- Deal with case of return from separate-session mobile app.
- F2F and DCMAW async both require similar logic and exception-handling. 
### Current discussion

- Users may have two equally-valid `CriResponseItems`. Which do we prioritise in making routing decisions? Either can be valid, failed or _pending_, which means we may need an assumption to simplify logic.

### NB

- Will squash when I merge, since there are a lot here.
- API tests will be covered in https://govukverify.atlassian.net/browse/PYIC-7785

### Issue tracking

- [PYIC-6246](https://govukverify.atlassian.net/browse/PYIC-6246)

[PYIC-6246]: https://govukverify.atlassian.net/browse/PYIC-6246?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Wynndow committed:
PYIC-6068: Remove writes to tactical in store-identity lambda

Removes the writing of VCs to the tactial store as well as EVCS in the
store identity lambda.

Wynndow committed:
PYIC-6068: Remove tactical writes from process-async lambda

Removes writing of VCs into the tactical store from the process async
CRI lambda.

Wynndow committed:
PYIC-6068: Remove ununsed env var from initialise session lambda

The initialise-ipv-session lambda already only writes to evcs, if the
flag is enabled. Writing to tactical must have already been removed. The
env var for the tactical store table name lingered on though and can be
binned.

MikeCollingwood committed:
pyic-6246: Centralise Async CRI status checking with new AsyncCriStatus class

MikeCollingwood committed:
pyic-6246: Add routing for incomplete identity but valid DCMAW Async VC

MikeCollingwood commented on libs/cri-response-service/src/main/java/uk/gov/di/ipv/core/library/service/CriResponseService.java:
```diff
@@ -62,15 +69,35 @@ public void persistCriResponse(
+        final boolean hasNoVc = !hasVc.apply(cri);
+        final boolean isComplete =
+                hasVc.apply(cri)
+                        && (!configService.enabled(EVCS_READ_ENABLED) || isPendingEvcsIdentity);
```
Assume we don't need to save/ consider other VCs from before strategic app, i.e. won't think about EVCS for dcmaw async.

Joe-Edwards-GDS replied:
(I haven't looked at any of the rest of this PR, or this particular line, but worth noting that we are _imminently_ removing this feature flag - https://govukverify.atlassian.net/browse/PYIC-6068 !)

MikeCollingwood committed:
pyic-6246: Update CheckMobileAppVcReceiptHandler for AsyncCriStatus changes

MikeCollingwood committed:
pyic-6246: Update abandon case for DCMAW async

MikeCollingwood committed:
pyic-6246: Convert AsyncCriStatus into a record

MikeCollingwood committed:
Merge branch 'main' into pyic-6246

MikeCollingwood committed:
pyic-6246: Update VerifiableCredentialBundle method names

MikeCollingwood committed:
pyic-6246: Re-introduce lombok to cri-response-service

MikeCollingwood committed:
pyic-6246: Fix CheckExistingIdentityHandlerTest unit tests

Wynndow committed:
PYIC-6068: Remove tactical reads from check-existing-identity

This removes reads of VCs from the tactical store from the
check-existing-identity lambda. It also removes code related to logging
differences between EVCS and tactical VCs, as well as the evcs migraiton
services.

Wynndow committed:
PYIC-6068: Remove tactical reads and writes from reset-session

This removes any use of the tactical store in the
reset-session-identity lambda.

Wynndow committed:
PYIC-6068: Remove unused env var from eval-gpg45

This must be a hangover from a previous iteration of the lambda, as it's
currently not used.

Wynndow committed:
PYIC-6068: Remove tactical reads from check-mobile-app-vc-receipt

This removes reads from the tactical store from the the
check-mobile-app-vc-receipt lambda and replaces it with calls to EVCS.

Wynndow committed:
PYIC-6068: Reorder status checks in check-mobile-app-vc-receipt

This change ensures we return early if the status is anything other than
pending. This means that we know we're returning a 404 when the VC is
missing AND the status is pending.

MikeCollingwood committed:
pyic-6246: Fix expired logic and add evcs vcs to session credentials if dcmaw async vc found

MikeCollingwood committed:
pyic-6246: Add unit tests for CheckExistingIdentityHandlerTest dcmaw async cri response

MikeCollingwood committed:
pyic-6246: Tidy up CheckExistingIdentityHandlerTest unit test

MikeCollingwood committed:
Merge branch 'main' into pyic-6246

MikeCollingwood committed:
pyic-6246: Fix case where F2F VC has no iat

MikeCollingwood committed:
pyic-6246: Remove check for EVCS feature flag for check existing identity in line with PYIC-6068

MikeCollingwood committed:
pyic-6246: Add clear check for DCMAW async vc iat not null

MikeCollingwood committed:
pyic-6246: Add unit tests and tidyup for sonar cloud complaints

MikeCollingwood committed:
pyic-6246: Make 0L default iat

MikeCollingwood committed:
pyic-6246: Remove iat default

MikeCollingwood committed:
pyic-6246: Add sonarcloud suppression for less specialised Function parameter

blakeyp commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -629,13 +632,40 @@ private Optional<JourneyResponse> checkForProfileMatch(
+            }
+
+            var isExpired =
+                    DateUtils.toSecondsSinceEpoch(new Date()) > asyncCriStatus.iat() + (30 * 60);
```
Probably be good to make this value configurable

blakeyp commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -629,13 +632,40 @@ private Optional<JourneyResponse> checkForProfileMatch(
+                    DateUtils.toSecondsSinceEpoch(new Date()) > asyncCriStatus.iat() + (30 * 60);
+            if (isExpired) {
+                LOGGER.info(LogHelper.buildLogMessage("DCMAW async VC expired."));
+                return JOURNEY_ERROR;
```
We probably want to kick off a new journey in this case (and for missing iat above?)

MikeCollingwood commented on libs/cri-response-service/src/main/java/uk/gov/di/ipv/core/library/service/CriResponseService.java:
```diff
@@ -62,15 +67,37 @@ public void persistCriResponse(
     }
+
+    @SuppressWarnings("java:S4276")
+    public AsyncCriStatus getAsyncResponseStatus(
```
There's a chance users can do both dcmaw async and f2f. Some discussed ideas were removing the CriResponseItem of an unsuccessful attempt (still unclear how we would handle pending state), OR we could try to prioritise CriResponses so we come to convenient conclusions for the user. Need product opinion

blakeyp replied:
Chatted with Mike - for the cross-browser recovery scenario we actually only care if we have a DCMAW Async VC issued in the last 30 minutes, we don't need to worry about the DCMAW pending record (we are assuming in vast majority of cases we'll have received the VC before the user logs back in). Suggest we check DCMAW (only for VC, not pending record) before we do F2F checks we do today

MikeCollingwood committed:
pyic-6246: Route users to new journey if expired DCMAW async and make ttl for expiry configurable

MikeCollingwood committed:
pyic-6246: Fix getNewIdentityJourney routes

MikeCollingwood committed:
pyic-6246: Update tests for new routes and oauthcriconfig fetch

Wynndow committed:
PYIC-6068: Use PENDING_RETURN for dcmaw async VC

The DCMAW async VC will be stored in EVCS as pending return. This
updates to correctly fetch it and store in in the session store.

MikeCollingwood committed:
pyic-6246: (BROKEN) Determine async CRI relevance with assumptions. Explain logic for readbility.

MikeCollingwood committed:
Merge branch 'pyic-6068-remove-reads-writes-to-from-tactical' into pyic-6246

MikeCollingwood committed:
pyic-6246: Initial fixes after merge

MikeCollingwood committed:
pyic-6246: Fix CheckExistingIdentityHandlerTest's

MikeCollingwood committed:
pyic-6246: Downgrade spotless

MikeCollingwood committed:
pyic-6246: Fix sonarcloud issues

MikeCollingwood commented on libs/cri-response-service/build.gradle:
```diff
@@ -7,6 +7,7 @@ plugins {
 
 dependencies {
 	implementation platform(libs.awsSdkBom),
+			libs.bundles.awsLambda,
```
This is for configService in CriResponseService

MikeCollingwood committed:
pyic-6246: Revert to isPendingReturn lingo

MikeCollingwood committed:
pyic-6246: Fix getAsyncResponseStatus stubbing test

MikeCollingwood committed:
pyic-6246: Revert rename of areGpg45VcsCorrelated

MikeCollingwood committed:
pyic-6246: Revert cosmetic changes to reduce PR mental load

MikeCollingwood committed:
pyic-6246: Revert cosmetic changes to reduce PR mental load

blakeyp commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -360,48 +381,27 @@ private VerifiableCredentialBundle getCredentialBundle(String userId, String evc
-            isPending = false;
+        if (isPendingReturn) {
+            // + vcs deferred from making an identity by an async CRI
+            evcsIdentityVcs.addAll(vcs.getOrDefault(PENDING_RETURN, List.of()));
```
Might be able to tidy this logic a little by just adding the inherited VCs here, otherwise return all? (if I'm following it correctly!)

blakeyp commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -248,13 +246,12 @@ private JourneyResponse getJourneyResponse(
+            var evcsAccessToken = clientOAuthSessionItem.getEvcsAccessToken();
+            var credentialBundle = getCredentialBundle(userId, evcsAccessToken);
+
+            var asyncCri =
```
`asyncCriStatus` ?

blakeyp commented on libs/cri-response-service/src/main/java/uk/gov/di/ipv/core/library/service/CriResponseService.java:
```diff
@@ -62,15 +71,45 @@ public void persistCriResponse(
+                        configService.getOauthCriConfigForConnection(connection, DCMAW_ASYNC);
+
+                var now = DateUtils.toSecondsSinceEpoch(new Date());
+                var expiry = DateUtils.toSecondsSinceEpoch(issuedAt) + criConfig.getTtl();
```
Hmm 'ttl' on CRI config might not be the best place for this, it's not a part of the CRI config 'standard' and this is really specific to cross-browser recovery, so maybe a root level config var something like `DCMAW_ASYNC_VC_PENDING_RETURN_TTL` ?

MikeCollingwood replied:
This makes sense. Have updated for this.

MikeCollingwood committed:
PYIC-6246: Use SSM parameter to define DCMAW VC ttl

MikeCollingwood committed:
PYIC-6246: Remove local dcmawAsync stub ttl

MikeCollingwood committed:
Merge branch 'main' into pyic-6246

MikeCollingwood committed:
Revert spotlessCheck omission from pre-commit config

MikeCollingwood committed:
PYIC-7785: Fix tests after update

MikeCollingwood committed:
PYIC-7785: Fix ErrorResponse code

MikeCollingwood committed:
PYIC-7785: Fix Sonarcloud issues

MikeCollingwood committed:
Merge branch 'main' into pyic-6246
